regulations/views/mixins.py

```diff
@@ -53,15 +56,39 @@ def build_sidebar_context(self, sidebar_class, label_id, version):
 
 
 class TableOfContentsMixin:
+    default_view = 'section_reader_view'
+
     def get_toc(self, reg_part, version):
         # table of contents
         toc = fetch_toc(reg_part, version)
         self.build_urls(toc, version)
         return toc
 
-    def build_urls(self, toc, version):
+    def build_urls(self, toc, version, subpart=None):
         for el in toc:
-            el['url'] = SectionUrl().of(
-                el['index'], version, self.sectional_links)
+            el['url'] = self.get_url(el['index'], version, subpart)
             if 'sub_toc' in el:
-                self.build_urls(el['sub_toc'], version)
+                if 'Subpart' in el['index']:
+                    self.build_urls(el['sub_toc'], version, '-'.join(el['index'][1:]))
+                else:
+                    self.build_urls(el['sub_toc'], version)
+
+    def get_url(self, citation, version, subpart):
+        view_name = self.request.resolver_match.url_name
+        part = citation[0]
+        section = citation[1]
+
+        if view_name == "subpart_reader_view" and subpart is not None:
+            view = view_name
+            args = (part, subpart, version)
+        elif view_name == "part_reader_view":
+            view = view_name
+            args = (part, version)
+        else:
+            view = self.default_view
+            args = (part, section, version)
```

Comment:
This conditional logic can move into the individual view classes now via an override of `get_url`